### ISSUE TITLE
NemoLite2D Fortran OpenMP optimisation

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSycloneBench#42]
In the manual implementation of the OpenMP Nemolite code, there exists a comment:
```
! We have to block here since sshn_t is used in the following loop.
! We could avoid this by altering the following two loop nests to read from
! ssha%data instead of sshn_t%data.
```

It seems trivial to resolve this to be able to add `NOWAIT` to this loop by accessing `ssha%data` in the following loops. The following loops only use data from the grid and (at the moment) `sshn_t%data`, which is updated by this loop (`sshn_t%data = ssha%data`).

I will look at benchmarking the current code on SKL with Intel and gcc, making this optimisation, and then rebenchmarking to see if it makes much difference, with schedule(static) and schedule(guided,X), with -O3.

**Comment by sergisiso** [comment]
Maybe it is time to make 2 versions of time_step_mod.f90 in psykal_omp (similarly of what psykal_serial and psykal_cpp do)? I think there is a point of having the code and performance of both:
- A naive OpenMP version (compare with PSyclone, what a non-expert may write, a maintainable code? ...)
- An optimized OpenMP version including the nowaits and this new optimization (compare with other PSyclone and other heavily optimized versions)

**Comment by LonelyCat124** [comment]
Current results, Skylake Node on scafellpike.
Commit hash: 2e2c999224db63241a2ff4740e1f0c56f6b0f651
All runs 3000 steps, irecord set to 3001 (no output in timed sections)
OMP_NUM_THREADS=32
Affinity not set.
Intel compiler - 19.3.199. 
CFLAGS = O3 -xCORE-AVX512 -qopt-report
F90FLAGS= `-O3 -xCORE-AVX512 -no-inline-min-size -no-inline-max-per-compile -no-inline-factor -qopenmp`

Schedule : STATIC
258x258: `0.95477`,`1.0394`,`0.96067`
1024x1024: `45.973`, `29.983`, `42.088` --some weirdness here?
2048x2048: `116.21`,`111.63`,`110.39`

Schedule: guided 32
258x258: `1.6526`,`2.1800`,`1.9539`
1024x1024: `26.188`, `34.084`, `35.133`
2048x2048: `112.59`, `113.41`, `111.61`

Schedule: guided 8
258x258: `1.7498`, `1.2131`, `1.1920`
1024x1024: `24.960`, 27.443`, `26.094`
2048x2048: `114.37`, `90.167`, 105.23`

------------------------------

gcc9.3:
CFLAGS: `-Ofast -mtune=native -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=skylake-avx512`
F90FLAGS: `-Wall -Wsurprising -Wuninitialized -march=skylake-avx512 -faggressive-function-elimination -Ofast -mtune=native -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=skylake-avx512 -ffree-line-length-none`

Schedule : STATIC
258x258: `13.356`, `12.754`, `12.730`
1024x1024: `83.848`, `83.584`, `83.642`
2048x2048: `229.11`, `225.63`, `221.22`

Schedule: guided 32
258x258: `9.2412`, `8.9740`, `9.1269`
1024x1024: `54.670`, `54.711`, `55.360`
2048x2048: `158.81`, `160.21`, `158.18`

Schedule: guided 8
258x258: `7.2367`, `7.1491`, `7.6511`
1024x1024: `53.328`, `52.775`, `53.426`
2048x2048: `164.42`, `167.62`, `162.71`


GCC performance is awful, on the smallcases over 10x for some comparisons! ARGH I LEFT IN THE MTUNE=NATIVE. Will redo them. Though even that doesn't explain 10x with static.

---
gcc 9.3
bash-4.2$ echo $CFLAGS
-Ofast -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=skylake-avx512
bash-4.2$ echo $F90FLAGS
-Wall -Wsurprising -Wuninitialized -march=skylake-avx512 -faggressive-function-elimination -Ofast -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=skylake-avx512 -ffree-line-length-none

Schedule : STATIC
258x258: `11.065`, `11.187`, `11.084`
1024x1024: `74.687`, `74.767`, `74.198`
2048x2048: `204.15`, `202.43`, `203.00`

Schedule: guided 32
258x258: `7.8659`, `8.5383`, `7.7564`
1024x1024: `49.340`, `48.930`, `49.097`
2048x2048:  `147.04`, `143.99`, `150.00`


Schedule: guided 8
258x258: `6.3039`, `6.4340`, `6.3152`
1024x1024: `47.924`, `49.519`, `48.242`
2048x2048:  `153.71`, `154.46`, `155.76`


gcc is still significantly slower but for the peak performance its only ~50%... @sergisiso any ideas?

**Comment by sergisiso** [comment]
Can you do a specific configuration (maybe guided 2048x2048) with `-fno-tree-vectorize / -novec` flags to see if the difference can be attributed to vectorization?

Also compiling with the `-mprefer-vector-width=512/-qopt-zmm-usage=high`  an also with a `avx2` arch could be another test?

**Comment by arporter** [comment]
I agree with @sergisiso's earlier comment about having a separate OMP version with these additional optimisations - it fits with what we've done for other versions. Any idea how these timings compare to what we had before?

**Comment by LonelyCat124** [comment]
@arporter These are all before changes timings for me to compare to after.
What would we be doing for the other openMP version? Just removing the nowaits?
I'll do those other pre benchmarks on 2048x2048 only for now @sergisiso 
I wish I'd written a script already!

**Comment by LonelyCat124** [comment]
gcc 9.3 with `-mprefer-vector-width=512`
schedule guided 32
2048x2048: `163.16`, `164.63`, `160.79`

gcc9 with `-fno-tree-vectorize`
schedule guided 32
2048x2048: `150.98`, `148.39`, `149.84`


intel 19 with `-qopt-zmm-usage=high`
schedule guided 32
2048:2048: `137,82` ,`138.35`, `140.02`

intel 19 with `novec`
schedule guided 32
2048x2048: `112.51`, `110.75`, `116.70`

**Comment by LonelyCat124** [comment]
New version, checksum is correct
gcc 9.3 with 
CFLAGS = `-Ofast -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=skylake-avx512`
F90FLAGS = `-Wall -Wsurprising -Wuninitialized -march=skylake-avx512 -faggressive-function-elimination -Ofast -finline-limit=50000 -fopt-info-all=gnu_opt_report.txt -march=skylake-avx512 -ffree-line-length-none`

schedule static:
258x258: `11.025`, `10.999`, `11.038`
1024x1024: `73.319`, `73.002`, `73.334`
2048x2048: `196.09`,`198.54`,`200.21`

schedule guided 32:
258x258: `7.7261`, `8.0715`, `7.6060`
1024x1024: `49.247`, `47.867` , `48.356`
2048x2048: `142.90`, `145.56`, `142.55`

schedule guided 8:
258x258: `6.3230`, `6.2497`, `6.2942`
1024x1024: `48.451`, `49.034`, `47.038`
2048x2048: `148.17`, `150.27`, `149.72`

intel 19.3
CFLAGS=`-O3 -xCORE-AVX512 -qopt-report`
F90FLAGS=`-O3 -xCORE-AVX512 -no-inline-min-size -no-inline-max-per-compile -no-inline-factor -qopt-report=5 -qopt-report-phase=loop,vec`

schedule static:
258x258: `1.2458`, `1.5806`, `1.3217`
1024x1024: `35.256`, `26.580`, `24.070`
2048x2048: `114.09`, `110.39`, `134.39`

schedule guided 32:
258x258:`2.5232`, `1.5342`, `1.8658`
1024x1024:`35.525`, `26.451`, `25.981`
2048x2048: `113.94`, `114.09`, `115.12`

schedule guided 8:
258x258: `1.2440`, `1.1670`, `2.5548`
1024x1024: `27.515`, `27.485`, `27.559`
2048x2048: `112.20`, `109.75`, `111.29`

**Comment by LonelyCat124** [comment]
Tough to argue if the best performance is better (intel with guided schedules is not noticably better), but the gcc results do seem better
